### DISCUSSION TITLE
Refreshing data from route data loader

### DISCUSSION THREAD

**Discussion by ngunna** [remix-run/react-router#10293] [Q&A] [answered]
I have data coming in on a route data loader and would like to retrigger the loader so I get fresh data in my component. As I see it I have a few options, none of which I particularly care for.

1. Trigger navigation. I have been able to get the results I want by simply using `useSearchParams` and setting search params even when there is nothing to set. This is will cause navigation and the loader to fire again giving me fresh data
2. Fetcher. At first glance this seemed like a good fit as it does precisely what I need, run a route loader again, however using them together starts to feel a bit cumbersome. The best way I have found to use them in conjunction is by storing loader data in state and watching `fetcher.data` for changes and updating state with updated fetcher data.
```
const loaderData = useLoaderData();
const fetcher = useFetcher();
const [items, setItems] = useState(loaderData);

useEffect(() => {
     setItems(fetcher.data)
}, [fetcher.data])
```
3. Avoid fetcher and simply fetch the data manually. You still end up needing local state to store the data, but at least fetching the data is slightly less of a pain. Since fetcher has issues with base path (https://github.com/remix-run/react-router/issues/9711) and only works off a full URL and cannot do any relative routing or some other simplified route matching (i.e. `useRouteLoaderData` which uses route id), getting at parent loader data is troublesome at the moment.

I apologize for the grumbling, the problem is solvable, but I couldn't help feeling like there has to be a better way. I am looking for any kind of guidance here, the docs don't quite cover the topic at the moment. I am hoping someone has solved this or there are plans to make this easier in the future. Perhaps something like `const [loaderData, triggerDataLoader] = useLoaderData()`, where `triggerDataLoader` is a function to re-run the data loader.

**Comment by kiliman** [comment, marked as answer]
Using loader for initial data, and fetchers for fresh data is a common pattern for Remix. I don't think it's too cumbersome. You can always create your own custom hook to wrap this.

**Reply by ngunna** [reply]
Thank you, I appreciate the response. I hope to see the fetcher basename fix (https://github.com/remix-run/react-router/issues/9711) and relative routing implemented (https://github.com/remix-run/react-router/discussions/9588) to help with usability. I will work around that stuff in the meantime.
